cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -85,76 +87,172 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
 
-      for (Map.Entry<String, MonitorConsumeRequest> entry : consumeRequests.entrySet()) {
-        String topicConfig = entry.getKey();
-        jsonWriter.name(topicConfig);
-        jsonWriter.beginArray();
+  /**
+   * A {@link BodyProducer} to encode and send back messages.
+   * Instead of using GenericDatumWriter, we perform the map and array encoding manually so that we don't have to buffer
+   * all messages in memory before sending out.
+   */
+  private static class MessagesBodyProducer extends BodyProducer {
+    private final Iterator<Map.Entry<String, GenericRecord>> requestsIterator;
+    // caches request key to topic
+    private final Map<String, String> requestKeyToLocalTopic;
+    private final MessageFetcher messageFetcher;
+    private final CConfiguration cConf;
+    private final Schema elementSchema = MonitorSchemas.V1.MonitorResponse.SCHEMA.getValueType().getElementType();
+    private final int numOfRequests;
+    private final int messageChunkSize;
+    private final Deque<GenericRecord> monitorMessages;
+    private final ByteBuf chunk;
+    private final Encoder encoder;
+    private final DatumWriter<GenericRecord> messageWriter;
+    private CloseableIterator<Message> iterator;
+    private boolean mapStarted;
+    private boolean mapEnded;
 
-        if (!requestKeyToLocalTopic.containsKey(topicConfig)) {
-          requestKeyToLocalTopic.put(topicConfig, getTopic(topicConfig));
+    MessagesBodyProducer(CConfiguration cConf, Map<String, GenericRecord> requests, MessageFetcher messageFetcher) {
+      this.requestsIterator = requests.entrySet().iterator();
+      this.requestKeyToLocalTopic = new HashMap<>();
+      this.messageFetcher = messageFetcher;
+      this.cConf = cConf;
+      this.numOfRequests = requests.size();
+      this.messageChunkSize = cConf.getInt(Constants.RuntimeMonitor.SERVER_CONSUME_CHUNK_SIZE);
+      this.monitorMessages = new LinkedList<>();
+      this.chunk = Unpooled.buffer(messageChunkSize);
+      this.encoder = EncoderFactory.get().directBinaryEncoder(new ByteBufOutputStream(chunk), null);
+      this.messageWriter = new GenericDatumWriter<GenericRecord>(elementSchema) {
+        @Override
+        protected void writeBytes(Object datum, Encoder out) throws IOException {
+          if (datum instanceof byte[]) {
+            out.writeBytes((byte[]) datum);
+          } else {
+            super.writeBytes(datum, out);
+          }
         }
+      };
+    }
+
+    @Override
+    public ByteBuf nextChunk() throws Exception {
+      // Already sent all messages, return empty to signal the end of response
+      if (mapEnded) {
+        return Unpooled.EMPTY_BUFFER;
+      }
+
+      chunk.clear();
+
+      if (!mapStarted) {
+        encoder.writeMapStart();
+        encoder.setItemCount(numOfRequests);
+        mapStarted = true;
+      }
+
+      // Try to buffer up to buffer size
+      int size = 0;
+
+      while (requestsIterator.hasNext() && size < messageChunkSize) {
```

Comment:
What happen if the `requestsIterator` has ended, but the `iterator` is not?

Reply:
Good catch!